### PR TITLE
No-batch-dim support for ConvNd

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70506

This PR adds no-batch-dim support for:
* `nn.Conv1d`, `nn.Conv2d`, `nn.Conv3d`
* `nn.LazyConv1d`, `nn.LazyConv2d`, `nn.LazyConv3d`
* `F.conv1d`, `F.conv2d`, `F.conv3d`
* `nn.ConvTranspose1d`, `nn.ConvTranspose2d`, `nn.ConvTranspose3d`
* `nn.LazyConvTranspose1d`, `nn.LazyConvTranspose2d`, `nn.LazyConvTranspose3d`
* `F.conv_transpose1d`, `F.conv_transpose2d`, `F.conv_transpose3d`

I tried to split these up into separate PRs but they use some common base classes so this broke tests :)

The support is added at the op level (e.g. `conv1d`) and tested via `ModuleInfo` entries with a reference function that compares the output from a single batch item input vs. the output from a no-batch-dim input.

Differential Revision: [D33355034](https://our.internmc.facebook.com/intern/diff/D33355034)

cc @ezyang @gchanan

### CODE REVIEW

**Comment by jbschlosser** [comment]
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

**Comment by jbschlosser** [comment]
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

File: test/test_modules.py [hunk 1]
```diff
@@ -417,9 +417,13 @@ def _test_gradients_helper(self, device, dtype, module_info, check):
 
             params = tuple(m.parameters())
 
-            # === Perform gradient check on the input_args ===
+            # === Lazy modules need to see an input to initialize params before gradcheck is run. ===
             input_args, input_kwargs = module_input.forward_input.args, module_input.forward_input.kwargs
+            if issubclass(module_info.module_cls, torch.nn.modules.lazy.LazyModuleMixin):
+                with torch.no_grad():
+                    m(*input_args, **input_kwargs)
```
Comment (jbschlosser) [inline test/test_modules.py:424]: Tweak needed here for lazy modules.

**Comment by jbschlosser** [comment]
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

File: aten/src/ATen/native/Convolution.cpp [hunk 2]
```diff
@@ -731,60 +764,87 @@ Tensor _convolution_mode(
...
-    const Tensor& input, const Tensor& weight, const c10::optional<Tensor>& bias_opt,
+    const Tensor& input_, const Tensor& weight, const c10::optional<Tensor>& bias_opt,
     IntArrayRef stride, IntArrayRef padding, IntArrayRef output_padding, int64_t groups, IntArrayRef dilation) {
   // See [Note: hacky wrapper removal for optional tensor]
   c10::MaybeOwned<Tensor> bias_maybe_owned = at::borrow_from_optional_tensor(bias_opt);
   const Tensor& bias = *bias_maybe_owned;
 
-  return at::convolution(input, weight, bias, stride, padding, dilation,
-                         true, output_padding, groups);
+  TORCH_CHECK(input_.dim() == 2 || input_.dim() == 3,
```
Comment (albanD) [inline aten/src/ATen/native/Convolution.cpp:809]: Why arent the transpose using the helper function?

File: torch/nn/modules/conv.py [hunk 3]
```diff
@@ -233,8 +233,8 @@ class Conv1d(_ConvNd):
     """.format(**reproducibility_notes, **convolution_notes) + r"""
 
     Shape:
-        - Input: :math:`(N, C_{in}, L_{in})`
-        - Output: :math:`(N, C_{out}, L_{out})` where
+        - Input: :math:`(N, C_{in}, L_{in})` or :math:`(C_{in}, L_{in})`
```
Comment (albanD) [inline torch/nn/modules/conv.py:236]: We don't have a syntax to say that the `N` dimension is optional?

File: torch/nn/modules/conv.py [hunk 4]
```diff
@@ -1203,6 +1209,12 @@ def __init__(
         if bias:
             self.bias = UninitializedParameter(**factory_kwargs)
 
+    def _get_in_channels(self, input: Tensor) -> int:
```
Comment (albanD) [inline torch/nn/modules/conv.py:1212]: nit: all of these are the same except the number of expected dimensions and function name. Should this be in a helper function like the c++ case?

File: torch/testing/_internal/common_modules.py [hunk 5]
```diff
@@ -314,6 +314,26 @@ def module_inputs_torch_nn_AvgPool1d(module_info, device, dtype, requires_grad,
                     reference_fn=no_batch_dim_reference_fn)]
 
 
+def module_inputs_torch_nn_ConvNd(module_info, device, dtype, requires_grad, **kwargs):
+    N = kwargs['N']
+    lazy = kwargs['lazy'] if 'lazy' in kwargs else False
```
Comment (albanD) [inline torch/testing/_internal/common_modules.py:319]: Here and below:
```python
lazy = kwargs.get('lazy', False)
```

**Review by albanD — COMMENTED** [review]
Looks mostly good.

File: torch/nn/modules/conv.py (hunk 3 above)
Comment (jbschlosser) [reply torch/nn/modules/conv.py:236]: We discussed using `[N]` but went forward with this throughout as it's more explicit without the need for a legend.

I could be convinced otherwise, but we'd need to make consistent changes across modules and add the syntax to [this guide](https://github.com/pytorch/pytorch/wiki/torch.nn-Module-Documentation-Style-Guide).

Imo this doesn't look too bad with a description inline:
``` Input: :math:`([N], C_{in}, L_{in})`, where :math:`[N]` is an optional batch dimension ```

File: aten/src/ATen/native/Convolution.cpp (hunk 2 above)
Comment (jbschlosser) [reply aten/src/ATen/native/Convolution.cpp:809]: Whoops, leftover from splitting the PR stack. Nice catch!

File: torch/nn/modules/conv.py (hunk 3 above)
Comment (albanD) [reply torch/nn/modules/conv.py:236]: I would personally vote for the `[N]` notation as it is shorter if we can use it consistently.

**Comment by jbschlosser** [comment]
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

**Comment by jbschlosser** [comment]
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

**Review by albanD — APPROVED** [review]
SGTM

lint needs to be fixed though

**Comment by jbschlosser** [comment]
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

**Comment by jbschlosser** [comment]
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

**Comment by jbschlosser** [comment]
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

**Comment by jbschlosser** [comment]
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

**Comment by jbschlosser** [comment]
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33355034).

**Comment by jbschlosser** [comment]
Labeled PR as BC-breaking because it was technically possible to (wrongly) pass e.g. 4D input / weight through `F.conv1d` before.
